src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Handlers/ProjectUpgradeHandler.cs

```diff
@@ -135,12 +137,16 @@ int IVsSolutionEvents.OnQueryUnloadProject(IVsHierarchy pRealHierarchy, ref int
         #endregion
 
 
-        [SuppressMessage("Microsoft.VisualStudio.Threading.Analyzers", "VSTHRD010", Justification = "NuGet/Home#4833 Baseline")]
         public void Dispose()
         {
             if (_cookie != 0 && _vsSolution2 != null)
             {
-                _vsSolution2.UnadviseSolutionEvents(_cookie);
+                NuGetUIThreadHelper.JoinableTaskFactory.RunAsync(async () =>
+                {
+                    await NuGetUIThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
+
+                    _vsSolution2.UnadviseSolutionEvents(_cookie);
+                });
```

Comment:
Always clear fields with cookies after unadvising. Dispose is supposed to be safe to call more than once. But your code here would unadvise twice, which because cookies get recycles could actually mean you're unadvising *someone else's* subscription.

Reply:
Well, you cleared the _vsSolution2 which I admit resolves the double-unadvise problem. But I'd personally be more comfortable if you cleared the cookie itself with:

```cs
_cookie = VSConstants.VSCOOKIE_NIL;
```

---
In reply to: [192149119](https://github.com/NuGet/NuGet.Client/pull/2269#discussion_r192149119) [](ancestors = 192149119)